compiler/rustc_hir_analysis/src/collect.rs

```diff
@@ -793,6 +793,10 @@ fn convert_enum_variant_types(tcx: TyCtxt<'_>, def_id: DefId) {
 fn find_field(tcx: TyCtxt<'_>, (def_id, ident): (DefId, Ident)) -> Option<FieldIdx> {
     tcx.adt_def(def_id).non_enum_variant().fields.iter_enumerated().find_map(|(idx, field)| {
         if field.is_unnamed() {
+            // Users can't write `expr._`, but derive macros can
+            if ident.name == kw::Underscore {
```

Comment:
This will always only return the first unnamed field, right?

Reply:
Yeah. I'm not really sure what a sensible way to handle all this is really. Maybe derives should just not work on _ fields?

Reply:
Considering that most user-written derive crates also try to look at field names, I think it'd make sense to ban `derive` for these kinds of ADTs altogether. Or maybe have a small list of allowed built-in derives (the [RFC](https://rust-lang.github.io/rfcs/2102-unnamed-fields.html#derive) says that deriving `Copy`, `Clone` or both should work).

Reply:
Ah, I missed that part of the RFC. Doing this on the Derive makes sense then (and maybe a `bug!()` if we get an underscore in `find_field`?)

Reply:
While we could add a `bug!` or `debug_assert_ne!` to `find_field` to catch faulty built-in macro impls, I don't know if it's wise to do so. `rustc_expand::…::expr_field` might be a better place.

I think you should first find the root cause of https://github.com/rust-lang/rust/issues/120992#issuecomment-1939731155 (doesn't involve a field access) & https://github.com/rust-lang/rust/issues/121263#issuecomment-1951154715 and fix it so that `$expr._` as generated by a HIR-based macro leads to an error instead of an ICE.

After that, probably in (a) separate PR(s) or in this PR I don't care, we could slowly add support for unnamed fields to built-in macros (and even if that means erroring out with *this trait cannot be derived on types with unnamed fields*) and finally add a `debug_assert_ne` to `expr_field`.

I don't think you can implement `#[derive(Clone)]` for these kinds of types yet since struct initialization doesn't support unnamed fields yet (we want to wait for that instead of using `MaybeUninit` lol).

Reply:
Makes sense. I'll see what I can do with that